File: frontend/apicostguard/src/services/appInfo.ts
import type { AppInfo } from "../types/domain";
import { tauriInvoke } from "./tauri";

function guessBrowserAppInfo(): AppInfo {
  const ua = navigator.userAgent;
  const platform = /Windows/i.test(ua)
    ? "windows"
    : /Mac/i.test(ua)
      ? "macos"
      : /Linux/i.test(ua)
        ? "linux"
        : "unknown";
  const arch = /arm64|aarch64|Apple A|arm;/i.test(ua) ? "aarch64" : /x64|WOW64|x86_64/i.test(ua) ? "x86_64" : "unknown";
  return { name: "APICostGuard", version: "dev", platform, arch };
}

export async function getAppInfo(): Promise<AppInfo> {
  try {
    return await tauriInvoke<AppInfo>("app_info");
  } catch {
    return guessBrowserAppInfo();
  }
}

File: frontend/apicostguard/src-tauri/src/commands/app_info.rs
use serde::Serialize;
use tauri::command;

#[derive(Serialize)]
pub struct AppInfo {
    pub name: String,
    pub version: String,
    pub platform: String,
    pub arch: String,
}

#[command]
pub fn app_info() -> AppInfo {
    AppInfo {
        name: "APICostGuard".into(),
        version: env!("CARGO_PKG_VERSION").into(),
        platform: std::env::consts::OS.into(),
        arch: std::env::consts::ARCH.into(),
    }
}
